compiler/src/dotty/tools/repl/ScriptEngine.scala

```diff
@@ -0,0 +1,90 @@
+package dotty.tools
+package repl
+
+import java.io.{Reader, StringWriter};
+import javax.script.{AbstractScriptEngine, Bindings, ScriptContext, ScriptEngine => JScriptEngine, ScriptEngineFactory, ScriptException, SimpleBindings}
+
+/** A JSR 223 (Scripting API) compatible wrapper around the REPL for improved
+ *  interoperability with software that supports it.
+ *
+ *  It works by instantiating a new script engine through the script engine manager.
+ *  The script engine provides a eval method to evaluate scripts in string form.
+ *  Example use:
+ *
+ *  val m = new javax.script.ScriptEngineManager()
+ *  val e = m.getEngineByName("scala")
+ *  println(e.eval("42"))
+ *
+ *  @param factory the script engine factory
+ */
+class ScriptEngine(factory: ScriptEngine.Factory) extends AbstractScriptEngine {
+  val driver = factory.driver
+  val rendering = new Rendering
+  var state: State = driver.initialState
+
+  def getFactory(): ScriptEngineFactory = factory
+
+  def createBindings: Bindings = new SimpleBindings
+
+  /* Evaluate with the given context. */
+  @throws[ScriptException]
+  def eval(script: String, context: ScriptContext): Object = {
+    val vid = state.valIndex
+    state = driver.run(script)(state)
+    val oid = state.objectIndex
+    Class.forName(s"rs$$line$$$oid", true, rendering.classLoader()(state.context))
+      .getDeclaredMethods.find(_.getName == s"res$vid")
+      .map(_.invoke(null))
+      .getOrElse(null)
+  }
+
+  @throws[ScriptException]
+  def eval(reader: Reader, context: ScriptContext): Object = eval(stringFromReader(reader), context)
+
+  def stringFromReader(reader: Reader) = {
+    val writer = new StringWriter()
```

Comment:
Be careful that it is StringWriter, not StringBuffer.

Reply:
Yes that's what I don't understand. Why use a `StringWriter` instead of a `StringBuilder`?

Reply:
Well, I take a writer because there is a reader. This is a trick I often use to pipe a reader to a writer. The right way to implement this method would be in a reactive manner, that is evaluate lines as and when they are coming in.

Reply:
> Well, I take a writer because there is a reader. This is a trick I often use to pipe a reader to a writer.

Even then, you can still use `StringBuilder`; the only point of `StringWriter` is if you need something to implement the `Writer` interface, which we don't. Otherwise, `StringBuilder.append(c.toChar)` works.

More importantly, reading *one character at a time* is pretty wasteful — that's what `read(char[] buf)` is for. This waste might be acceptable here.

> The right way to implement this method would be in a reactive manner, that is evaluate lines as and when they are coming in.

Indeed, this loop reads until the end of the stream, that seems problematic as you say. What about testing `reader.ready()` instead?

Reply:
Yes, it might be relevant. However, I would like to point out that this method is seldom used, in Scala or other scripting languages. This is a quick hack to implement the ScriptEngine interface. We could also throw UnsupportedOperationException. If a relevant use case comes, we can always implement it in a clean (and tested) fashion. For the moment, my goal is not to go further than 2.12. What do you think ?

Reply:
> This is a quick hack to implement the ScriptEngine interface. We could also throw UnsupportedOperationException.

👍